src/common/hooks/account/use-api-nonce.ts

```diff
@@ -0,0 +1,36 @@
+import { lastApiNonceAtom } from '@store/accounts/nonce';
+import { useUpdateAtom } from 'jotai/utils';
+import { apiClientState } from '@store/common/api-clients';
+import { currentAccountStxAddressState } from '@store/accounts';
+import { useQuery } from 'react-query';
+import { useAtomValue } from 'jotai/utils';
+import { AddressNonces } from '@stacks/blockchain-api-client/lib/generated';
+
+export const correctNextNonce = (apiNonce: AddressNonces): number | undefined => {
+  if (!apiNonce) return;
+
+  const missingNonces = apiNonce.detected_missing_nonces;
+  if (
+    missingNonces &&
+    missingNonces.length > 0 &&
+    missingNonces[0] > (apiNonce.last_executed_tx_nonce || 0) &&
+    missingNonces[0] > (apiNonce.last_mempool_tx_nonce || 0)
+  ) {
+    return missingNonces[0];
+  }
+  return apiNonce.possible_next_nonce;
+};
+
+export function useApiNonce() {
+  const setLastApiNonceAtom = useUpdateAtom(lastApiNonceAtom);
+  const principal = useAtomValue(currentAccountStxAddressState);
+  if (!principal) return;
+  const { accountsApi } = useAtomValue(apiClientState);
+  const fetchNonce = () => accountsApi.getAccountNonces({ principal });
+
+  const { data: apiNonce } = useQuery(['apiNonce'], fetchNonce);
+  if (!apiNonce) return;
+  const nextNonce = correctNextNonce(apiNonce);
+  if (typeof nextNonce === 'number') setLastApiNonceAtom(nextNonce);
+  return apiNonce;
+}
```

Comment:
A few comments on this:
- you should never have a component that can be rendered with a change in the number of hooks. Currently because of L27, it's possible the `useQuery` hook can be rendered and then removed. This breaks the[ rules of hooks.](https://reactjs.org/docs/hooks-rules.html)
- L34 you're setting a state during the render cycle, this means it will be set every time it renders. You'd either do this in a useEffect or with `onSuccess`
```suggestion
export function useApiNonce() {
    const [nonce, setLastApiNonceAtom] = useAtom(lastApiNonceAtom);
    const principal = useAtomValue(currentAccountStxAddressState);
    const { accountsApi } = useAtomValue(apiClientState);
    const fetchNonce = () => accountsApi.getAccountNonces({ principal });

    useQuery(['apiNonce', principal], fetchNonce, {
        enabled: !!principal,
        onSuccess: (data) => {
            const nextNonce = correctNextNonce(data);
            if (typeof nextNonce === 'number') setLastApiNonceAtom(nextNonce);
        }
    });
    
    return nonce;
}
```

Reply:
I'd also do this:
```ts
    const network = useAtomValue(currentNetworkState);
    useQuery(['apiNonce', principal, network.url], fetchNonce, {
```

Reply:
and just as a point of reference, here is the same `useQuery` hook, but as an atom:
```ts
const apiNonceAtom = atom(get => {
  const principal = get(currentAccountStxAddressState);
  const network = get(currentNetworkState);

  const anAtom = atomWithQuery(
    ['apiNonce', principal, network.url],
    () => {
      const { accountsApi } = get(apiClientState);
      if (!principal) return;
      return accountsApi.getAccountNonces({
        principal,
      });
    },
    {
      enabled: !!principal,
    }
  );

  return get(anAtom);
});
```

Reply:
This should have caused a lint error with the react rules of hook eslint plugin 🤔

Reply:
it's more subtle than that, not sure if the eslint plugin can catch it -- and will throw if a component is rendered with a different number of hooks.

Reply:
It should be able to detect conditional use of hooks
![image](https://user-images.githubusercontent.com/1618764/130984685-1405f13b-8dbe-42b5-b095-97744b2beb73.png)

Reply:
oh nice! didn't know it could

Reply:
Thanks for the detailed review ! 
I think this is missing in the eslint.rc
 "react-hooks/rules-of-hooks": "error",
When I added it it shows the warning after. I will check and add it if it makes sense.